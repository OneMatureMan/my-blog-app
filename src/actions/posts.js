import { v1 as uuidv1 } from 'uuid';
import moment from 'moment';

export const addPost = (
    {
        title = '',
        body = '',
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_POST',
    post : {
        id: uuidv1(),
        title,
        body,
        createdAt: moment()
    }
})

export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});


export const removePost = ({ id } = {}) => ({
    type: 'REMOVE_POST',
    id
})