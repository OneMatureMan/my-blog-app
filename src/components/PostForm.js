import React from 'react';


class PostForm extends React.Component {
    state = {
        title: this.props.post ? this.props.post.title : '',
        body: this.props.post ? this.props.post.body : '',
        error: ''
    }


    handleTitleChange = (e) => {
        const title = e.target.value;
        this.setState(({ title }))
    }
    
    handleBodyChange = (e) => {
        const body = e.target.value;
        this.setState(({ body }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.body) {
            return this.setState(({ error: 'Please enter both title and body'}))
        }
        this.setState(({ error: '' }))
        this.props.onSubmit({
            title: this.state.title,
            body: this.state.body
        })        
    }
    
    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                    type="text"
                    name='title'
                    placeholder='Title'
                    autoFocus
                    value = {this.state.title}
                    onChange={this.handleTitleChange}
                />
                <textarea
                    name='body'
                    placeholder='Enter the blogs body here'
                    value= {this.state.body}
                    onChange={this.handleBodyChange}
                />
                <button>Save Post</button>
            </form>
        )
    }
}


export default PostForm;