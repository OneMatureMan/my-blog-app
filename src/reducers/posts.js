const postsDefaultState = []

const postsReducer = (state = postsDefaultState, action) => {
    switch (action.type) {
        case 'ADD_POST':
            return state.concat({ id:action.id, ...action.post});
        case 'EDIT_POST':
            return state.map(post => {
                if (post.id === action.id) {
                    return {
                        ...post,
                        ...action.updates
                    }
                } else {
                    return post;
                }
            })
        case 'REMOVE_POST':
            return state.filter(({ id }) => id !== action.id)
        default :
            return state;
    }
}

export default postsReducer;