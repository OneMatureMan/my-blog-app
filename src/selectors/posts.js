// get visible posts
export default (posts, { text, sortBy }) => {
    return posts.filter(({ title }) => title.toLowerCase().includes(text.toLowerCase()))
            .sort((a,b) => {
                if (sortBy === 'date') {
                    if (a.createdAt < b.createdAt) {
                        return 1;
                    }
                    if (a.createdAt > b.createdAt) {
                        return -1;
                    }
                    return 0;
                } else if (sortBy === 'title') {
                    
                    if (a.title.toLowerCase() < b.title.toLowerCase()) {
                        return -1;
                    }
                    
                    if (a.title.toLowerCase() > b.title.toLowerCase()) {
                        return 1;
                    }
                    return 0;
                }            
            })
}