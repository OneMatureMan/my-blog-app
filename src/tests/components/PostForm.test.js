import React from 'react';
import { shallow } from 'enzyme';
import PostForm from '../../components/PostForm';
import posts from '../fixtures/posts';

test('should render Post form component correctly', () => {
    const wrapper = shallow(<PostForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should submit form with data', () => {
    const onSubmit = jest.fn();
    const wrapper = shallow(<PostForm post={posts[0]} onSubmit={onSubmit} />);
    wrapper.find('form').simulate('submit', {
        preventDefault : () => {}
    });
    expect(onSubmit).toHaveBeenCalledWith({title:posts[0].title, body:posts[0].body})
});


test('should set title on change', () => {
    const value = 'title'
    const wrapper = shallow(<PostForm />);
    wrapper.find('input').simulate('change', {
        target : { value }
    });
    expect(wrapper.state('title')).toBe('title')
});

test('should set body on change', () => {
    const value = 'body'
    const wrapper = shallow(<PostForm />);
    wrapper.find('textarea').simulate('change', {
        target : { value }
    });
    expect(wrapper.state('body')).toBe('body');
});