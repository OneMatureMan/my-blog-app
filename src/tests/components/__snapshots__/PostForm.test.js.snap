// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Post form component correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    name="title"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    name="body"
    onChange={[Function]}
    placeholder="Enter the blogs body here"
    value=""
  />
  <button>
    Save Post
  </button>
</form>
`;
